compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1454,6 +1498,20 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
                   fourthTry
           || tryLiftedToThis2
 
+          if isEffCheckingOrSetup && onlyEffCheckKill then
+              // if expected type is not dependent function type e.g. File^ => Unit
+              // and actual type is dependent function type with kill type e.g. (f: File^) => Unit @kill(f)
+              // then it should fail - in this case expected type is just an AppliedType
```

Comment:
To further document the motivation here:

> This is because in Scala dependent function types are automatically subtypes of non-dependent function types (which are represented as AppliedTypes).
So by default in Scala something like
`(f: File^ => Unit @kill(f)) <: File^ => Unit`
is true. So this is to check for that case.